homeassistant/components/overkiz/__init__.py

```diff
@@ -43,17 +54,58 @@ class HomeAssistantOverkizData:
     scenarios: list[Scenario]
 
 
+async def async_migrate_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Migrate an old config entry."""
+
+    LOGGER.debug("Migrating from version %s", entry.version)
+
+    # v1 -> v2: CONF_HUB renamed to CONF_SERVER and CONF_API_TYPE added
+    if entry.version == 1:
+        v2_entry_data = {**entry.data}
+        v2_entry_data[CONF_SERVER] = entry.data[CONF_HUB]
+        v2_entry_data.pop(CONF_HUB)
+        v2_entry_data[CONF_API_TYPE] = APIType.CLOUD  # V1 only supports cloud
+
+        entry.version = 2
+        hass.config_entries.async_update_entry(entry, data=v2_entry_data)
+
+        LOGGER.debug("Migration to version %s successful", entry.version)
+
+    return True
+
+
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Overkiz from a config entry."""
-    username = entry.data[CONF_USERNAME]
-    password = entry.data[CONF_PASSWORD]
-    server = SUPPORTED_SERVERS[entry.data[CONF_HUB]]
-
-    # To allow users with multiple accounts/hubs, we create a new session so they have separate cookies
-    session = async_create_clientsession(hass)
-    client = OverkizClient(
-        username=username, password=password, session=session, server=server
-    )
+
+    client: OverkizClient | None = None
+
+    # Local API
+    if entry.data[CONF_API_TYPE] == APIType.LOCAL:
+        host = entry.data[CONF_HOST]
+        token = entry.data[CONF_TOKEN]
+
+        # Verify SSL blocked by https://github.com/Somfy-Developer/Somfy-TaHoma-Developer-Mode/issues/5
+        # Somfy (self-signed) SSL cert uses the wrong common name
+        session = async_create_clientsession(hass, verify_ssl=False)
+
+        client = OverkizClient(
+            username="",
+            password="",
+            token=token,
+            session=session,
+            server=generate_local_server(host=host),
+        )
+    # Overkiz Cloud API
+    else:
+        username = entry.data[CONF_USERNAME]
+        password = entry.data[CONF_PASSWORD]
+        server = SUPPORTED_SERVERS[entry.data[CONF_SERVER]]
+
+        # To allow users with multiple accounts/hubs, we create a new session so they have separate cookies
+        session = async_create_clientsession(hass)
+        client = OverkizClient(
+            username=username, password=password, session=session, server=server
+        )
```

Comment:
This might be a bit easier to maintain if you split this into a local and remote function to get the client